modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+    gemm(A, B, C, 1:eltType, 0:eltType, opA=Op.T);
+    return C;
+  }
+}
+
+
+/* Transpose vector or matrix */
+proc _array.T where this.domain.rank == 2 { return transpose(this); }
+
+
+/* Transpose matrix with native implementation */
+pragma "no doc"
+private proc _transpose(A: [?Dom] ?eltType) {
+  var C: [transpose(Dom)] eltType;
+
+  // naive algorithm
+  forall (i, j) in Dom do
+    C[j, i] = A[i, j];
+
+  return C;
+}
+
+
+/* Add matrices, maintaining dimensions */
+proc matPlus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Ranks to not match");
+  if Adom.size != Bdom.size then halt("Unmatched array size");
+  var C: [Adom] eltType = A + B;
+  return C;
+}
+
+
+/* Subtract matrices, maintaining dimensions */
+proc matMinus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != Bdom.rank then compilerError("Ranks to not match");
+  if Adom.size != Bdom.size then halt("Unmatched array size");
+  var C: [Adom] eltType = A - B;
+  return C;
+}
+
+/*
+    Generic matrix multiplication, ``A`` and ``B`` can be a matrix, vector, or
+    scalar
+*/
+proc dot(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  // vector-vector
+  if Adom.rank == 1 && Bdom.rank == 1 then
+    return inner(A, B);
+  // matrix-(vector|matrix)
+  else
+    return matMult(A, B);
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(A: [?Adom] ?eltType, b) where isNumeric(b) {
+  var C: [Adom] eltType = A*b;
+  return C;
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(a, B: []) where isNumeric(a) {
+  return dot(B, a);
+}
+
+
+pragma "no doc"
+/* Explicit matrix-(matrix|vector) multiplication */
+proc matMult(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  // matrix-vector
+  if Adom.rank == 2 && Bdom.rank == 1 then
+    // TODO -- assert shapes are correct
+    return _matvecMult(A, B);
+  // vector-matrix
+  else if Adom.rank == 1 && Bdom.rank == 2 then
+    // TODO -- assert shapes are correct
+    return _matvecMult(B, A);
+  // matrix-matrix
+  else if Adom.rank == 2 && Bdom.rank == 2 then
+    // TODO -- assert shapes are correct
+    return _matmatMult(A, B);
+  else
+    compilerError("Rank sizes are not 1 or 2");
+}
+
+
+pragma "no doc"
+/* matrix-vector multiplication */
+private proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Xdom.rank != 1 then
+    compilerError("Rank sizes are not 2 and 1");
+
+  var Y: [Xdom] eltType;
+  gemv(A, X, Y, 1:eltType, 0:eltType);
+  return Y;
+}
+
+
+pragma "no doc"
+/* matrix-matrix multiplication */
+private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Bdom.rank != 2 then
+    compilerError("Rank sizes are not 2");
+
+  var C: [Adom.dim(1), Bdom.dim(2)] eltType;
+  gemm(A, B, C, 1:eltType, 0:eltType);
+  return C;
+}
+
+
+/* Inner product of 2 vectors */
```

Comment:
It is redundant with `dot`, but I believe that argument here (following numpy's interface), is that it can be useful to explcitly distinguish `inner` from `outer` products. We could also later extend `inner` to support higher-dimensional arrays, like numpy.